HeartbeatLogging/Tests/Integration/HeartbeatLoggingIntegrationTests.swift

```diff
@@ -13,16 +13,307 @@
+    // When
+    heartbeatController.log("dummy_agent")
+    // Then
+    assertHeartbeatControllerFlushesEmptyPayload(heartbeatController)
+  }
+
+  func testMultipleControllersWithTheSameIDUseTheSameStorageInstance() throws {
+    // Given
+    let heartbeatController1 = HeartbeatController(id: #function, dateProvider: { self.date })
+    let heartbeatController2 = HeartbeatController(id: #function, dateProvider: { self.date })
+    // When
+    heartbeatController1.log("dummy_agent")
+    // Then
+    let payload = heartbeatController2.flush()
+    try assertEqualPayloadStrings(
+      payload.headerValue(),
+      """
+      {
+        "version": 2,
+        "heartbeats": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
+    assertHeartbeatControllerFlushesEmptyPayload(heartbeatController1)
+  }
+
+  func testHeartbeatControllerConcurrencyStressTest() throws {
+    // Given
+    let heartbeatController = HeartbeatController(id: #function, dateProvider: { self.date })
+
+    // When
+    DispatchQueue.concurrentPerform(iterations: 100) { _ in
+      heartbeatController.log("dummy_agent")
+    }
+
+    var payloads: [HeartbeatsPayload] = []
+
+    DispatchQueue.concurrentPerform(iterations: 100) { _ in
+      let payload = heartbeatController.flush()
+      payloads.append(payload)
+    }
+
+    // Then
+    let nonEmptyPayloads = payloads.filter { payload in
+      payload.headerValue().count > 0
+    }
+
+    XCTAssertEqual(nonEmptyPayloads.count, 1)
+    let payload = try XCTUnwrap(nonEmptyPayloads.first)
+    try assertEqualPayloadStrings(
+      payload.headerValue(),
+      """
+      {
+        "version": 2,
+        "heartbeats": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
   }
+
+  func testLogRepeatedly_WithoutFlushing_LimitsOnWrite() throws {
+    // Given
+    var testdate = date
+    let heartbeatController = HeartbeatController(id: #function, dateProvider: { testdate })
+
+    // When
+    // Iterate over 35 days and log a heartbeat each day.
+    // - 30: The heartbeat logging library can store a max of 30 heartbeats. See
+    //   `HeartbeatController`'s `heartbeatsStorageCapacity` property.
+    // - 5: Because of the above limit, expect 5 heartbeats to be overwritten.
+    for day in 1 ... 35 {
+      // A different user agent is logged based on the current iteration. There
+      // is no particular reason for when each user agent is used– the goal is
+      // to achieve a payload with multiple user agent groupings.
+      if day < 5 {
+        heartbeatController.log("dummy_agent_1")
+      } else if day < 13 {
+        heartbeatController.log("dummy_agent_2")
+      } else {
+        heartbeatController.log("dummy_agent_3")
+      }
+
+      testdate.addTimeInterval(60 * 60 * 24) // Advance the test date by 1 day.
```

Comment:
No, it shouldn't because the whatever is passed in as the `dateProvider` is "standardized" with a configured `Calendar`.  

https://github.com/firebase/firebase-ios-sdk/blob/10410ec6cc5763fe55dfe941f366008b7abb3952/HeartbeatLogging/Sources/HeartbeatController.swift#L52-L60

The standardizer is statically configured at the top of class: 

https://github.com/firebase/firebase-ios-sdk/blob/10410ec6cc5763fe55dfe941f366008b7abb3952/HeartbeatLogging/Sources/HeartbeatController.swift#L26-L33

So this ensures a standardized date is ultimately recorded that isn't biased by factors like daylight savings.

FWIW, when I first wrote this test, it did fail due to DLS because Nov 7 (i think?) was daylight savings time so the payload string one day off and i couldn't figure out why for the life of me. It was because the `HeartbeatController` date was not standardized to ignore factors like DSL so adding `60 * 60 * 24` resulted in the same day on the day of DLS.